### ISSUE TITLE
Incorrect return type declaration in Post::determine_id()

### ISSUE THREAD

**Issue by FlyingDR** [timber/timber#2184]
At a time of writing this bug report both [1.x](https://github.com/timber/timber/blob/828bdcc8bfaeb4b76b49d4dba8baec3ea6950a8e/lib/Post.php#L228) and [2.x](https://github.com/timber/timber/blob/03bf4bf20dc763895b776b7f51ef856050821da8/lib/Post.php#L303) branches of Timber declares return type of `\Timber\Post::determine_id()` as `int` while actual return type is `int|null` because it is possible to have no post at all, e.g. inside of `404.php`.

Since `\Timber\Post` accepts `null` for `$pid` in its constructor - returning `null` from `determine_id()` effectively propagates this (unexpected) value further that may lead to unexpected behavior.

Additionally, if Timber, which requires `^7.0` for its 2.x branch will include language level declaration of return types - this inconsistency will produce [`\TypeError`](https://www.php.net/manual/en/class.typeerror.php) error.

**Comment by jarednova** [comment]
Thanks @FlyingDR. What's the shortest path to resolve? Is it as simple as updating the docblock for `determine_id()`? 🤞. Something to do with implementing #1863? Or does this require a deeper voyage in order to successfully root out?

**Comment by gchtr** [comment]
I think @acobster will sort this out in his work on factories in the `2.x-factories` branch. I’m sorry I don’t have any specific pull request or issue to show you, because this is still all in the works.

I don’t think we are going to solve this in Timber 1.x, because issues like these are one reason we started working on Timber 2.x.

> Since `\Timber\Post` accepts `null` for `$pid` in its constructor - returning `null` from `determine_id()` effectively propagates this (unexpected) value further that may lead to unexpected behavior.

This is exactly something we want to prevent with factories behind the scenes and the new API (#2073).

I think it’s good you raised this issue, because we can add specific tests for this. 👍 It’s probably best to let this rest for now and come back to it when we work out the details for post factories.

**Comment by FlyingDR** [comment]
> Thanks @FlyingDR. What's the shortest path to resolve? Is it as simple as updating the docblock for `determine_id()`? crossed_fingers. Something to do with implementing #1863? Or does this require a deeper voyage in order to successfully root out?

I think for 1.x it will be enough to just update PHPDoc to give developers proper information about actual types. For 2.x you probably know better what to do.

**Comment by acobster** [comment]
@FlyingDR yeah the design of Factories pulls apart a few responsibilities that currently are tangled up inside `determine_id()`. My hope is that we can completely discard that method in 2.x.

As @gchtr mentioned you can look at the Factory classes on [2.x-factories](https://github.com/timber/timber/tree/2.x-factories/lib/Factory). This design as a whole is still a work in progress but the Factories themselves are very close to what will ship in 2.0.

In short:

* Any logic that relies on global state, e.g. `get_the_ID()` goes in the top-level `Timber::get_*()` methods and is passed down to a Factory.
* Once a Factory has an ID/post/query/etc. to work with, it converts that into a core WP object, e.g. `WP_Post`.

**Comment by FlyingDR** [comment]
@acobster Thank you for reference. I've look over the code and noticed some minor issues, I will add this information as comments to #2093 so it will not get lost.

**Comment by acobster** [comment]
This method doesn't exist anymore. Its responsibilities were taken over by `PostFactory` in 2.0, and the semantics are now much better defined. In short, if `Timber::get_post()` doesn't find a post, it returns `false`. The upgrade guide will be updated soon with all the relevant info.
